cpp/oneapi/dal/algo/pca/backend/gpu/train_kernel_cov_dpc.cpp

```diff
@@ -38,107 +34,85 @@ using result_t = train_result<task::dim_reduction>;
-    const auto daal_eigvec =
-        interop::convert_to_daal_table(queue, arr_eigvec, component_count, column_count);
-    const auto daal_eigval = interop::convert_to_daal_table(queue, arr_eigval, 1, component_count);
-    const auto daal_means = interop::convert_to_daal_table(queue, arr_means, 1, column_count);
-    const auto daal_variances = interop::convert_to_daal_table(queue, arr_vars, 1, column_count);
-
-    daal_cov::Batch<Float, daal_cov::defaultDense> covariance_alg;
-    covariance_alg.input.set(daal_cov::data, daal_data);
-
-    constexpr bool is_correlation = false;
-    constexpr std::uint64_t results_to_compute =
-        std::uint64_t(daal_pca::mean | daal_pca::variance | daal_pca::eigenvalue);
-    const bool is_deterministic = desc.get_deterministic();
-
-    call_daal_gpu_kernel<Float>(is_correlation,
-                                is_deterministic,
-                                *daal_data,
-                                &covariance_alg,
-                                static_cast<DAAL_UINT64>(results_to_compute),
-                                *daal_eigvec,
-                                *daal_eigval,
-                                *daal_means,
-                                *daal_variances);
-
-    // clang-format off
-    const auto model = model_t{}
-        .set_eigenvectors(
-            dal::detail::homogen_table_builder{}
-                .reset(arr_eigvec, component_count, column_count)
-                .build()
-        );
+    const auto data_nd = pr::flatten_table<Float, row_accessor>(q, data, sycl::usm::alloc::device);
+
+    auto [sums, sums_event] = compute_sums(q, data_nd);
+    auto [corr, means, vars, corr_event] = compute_correlation(q, data_nd, sums, { sums_event });
+
+    auto [eigvecs, eigvals] =
+        compute_eigenvectors_on_host(q, std::move(corr), component_count, { corr_event });
+
+    if (desc.get_deterministic()) {
+        sign_flip(eigvecs);
+    }
+
+    const auto model = model_t{}.set_eigenvectors(
+        homogen_table::wrap(eigvecs.flatten(), component_count, column_count));
 
     return result_t{}
         .set_model(model)
-        .set_eigenvalues(
-            dal::detail::homogen_table_builder{}
-                .reset(arr_eigval, 1, component_count)
-                .build()
-        )
-        .set_variances(
-            dal::detail::homogen_table_builder{}
-                .reset(arr_vars, 1, column_count)
-                .build()
-        )
-        .set_means(
-            dal::detail::homogen_table_builder{}
-                .reset(arr_means, 1, column_count)
-                .build()
-        );
-    // clang-format on
-}
-
-template <typename Float>
-static result_t train(const context_gpu& ctx, const descriptor_t& desc, const input_t& input) {
-    return call_daal_kernel<Float>(ctx, desc, input.get_data());
+        .set_eigenvalues(homogen_table::wrap(eigvals.flatten(), 1, component_count))
```

Comment:
I think, keeping data on host is a better strategy, we don't know how the data is accessed after PCA. If it's accessed by CPU algo, no extra data copy is required. If it's accessed by GPU, the copy will be done once table is accessed by `accessor` with the proper queue.

In C++ interface we don't have problem like in scikit-learn extension as data storage and access logic are decoupled. We don't even guarantee that it will be always homogen table.

Reply:
>I think, keeping data on host is a better strategy

Only in this case because here we compute eigenvecs on CPU. Here I think might be two possible strategies:
1. To return the resulting data on the location where it was computed - most performant option I guess since it helps to avoid unnecessary data transfers
2. To return the data on the device where an execution was triggered by the user.

Maybe the first option is better until we indicate the data location of results in our docs.